### DISCUSSION TITLE
http header Variable auslesen?

### DISCUSSION THREAD

**Discussion by faulgurke** [e-netze/datalinq-community#19] [Q&A]
Wie kann ich eine header Variable wie z.B. "X-Authenticate-UserID" in eine razor Variable übernehmen?

**Comment by gis-eni** [comment]
Die Razor Komponente kann nicht auf den Request zugreifen. Somit auch nicht auf die Header Variablen.
Vielleicht beschreibst du einmal deine Anforderungen, vielleicht gibt es dazu einen Workaround.

**Reply by ue-black** [reply]
Zum Hintergrund:
Wir arbeiten daran, dass eine Fachabteilung den Inhalt einer Datalinq Anwendung ("Badestellen") selbst editieren kann. Konkret geht es darum einen aktuellen Info-Text zu schalten. Dafür haben wir einen eigenen Endpoint gebaut wo wir mittels Formular Daten in unsere DB schreiben, die wiederum dann vom Endpoint "Badestellen" ausgelesen wird. Das ganze läuft im Portalverbund und wir möchten gerne den Header auslesen und in die DB schreiben damit wir wissen wer das erfasst hat.

So schaut die Liste an Meldungen die man über unser primitives Formular eingeben kann aus (im Testsystem):
![image](https://github.com/user-attachments/assets/88af9f87-6df4-4c7b-b05f-e4c10283ee8a)
 
Das ist die produktive Anwendung (der Infotext ändert sich alle zwei Wochen im Sommer): https://atlas.vorarlberg.at/badegewaesser

**Reply by gis-eni** [reply]
Ok, wenn du den aktuellen Usernamen willst kannst du das über das Model des Razor View abfragen:

``` csharp
Model.UserInformation.Username
```

Zusätzlich könntest du über die ``UserInformation`` (Type: ``IDataLinqUser``) auch auf die Rollen des Users zugreifen;

``` csharp
public interface IDataLinqUser
{
    string Username { get; }

    IEnumerable<string> Userroles { get; }
}
```

Schau einmal, ob der Username, den du da bekommst, das ist was du brauchst.

**Reply by faulgurke** [reply]
Mit der ersten Variante bekomme mit folgendem Code gleich beim Speichern eine Fehlermeldung:
@{ var user = Model.UserInformation.Username;}
<table id = "anzahl">
    <tr>
        <td>Benutzer:  @user </td>
    </tr>
</table>
Fehler:  'SelectResult' does not contain a definition for 'UserInformation' and no accessible extension method 'UserInformation' accepting a first argument of type 'SelectResult' could be found (are you missing a using directive or an assembly reference?

MIt der zweiten Variante (UserInformation und Zugriff auf Rollen) schlägt mir die KI die Verwendung von "Microsoft" und "System" Libraries vor, was ja nicht geht:
using Microsoft.AspNetCore.Mvc.RazorPages;
using System.Collections.Generic;

Vielleicht kannst du deinen Vorschlag mit IDataLinqUser detaillierter ausführen?

**Reply by gis-eni** [reply]
Du hast recht. ``UserInformation`` ist ``private``. Darum kann man darauf nicht zugreifen. Das wurde nur umgesetzt, damit man die *DataLinqHelper* Methode ``@DLH.HasRole(...)`` verwenden kann.

Einen direkten Zugriff den Usernamen hat man somit nicht. Hierzu müsste man DataLinq erweitern. Neu Features kommen allerdings nur mehr in die Version von WebGIS6... Verwendet ihr das schon?

**Reply by faulgurke** [reply]
Nein, noch nicht. Aber das ist ein seeehr guter Grund, sich das ernsthaft zu überlegen ;-)

**Comment by gis-eni** [comment]
In neuen Datalinq/WebGIS Releases (>=6.24.3601) kommen neue DataLinqHelper:

``` csharp
@DLH.GetCurrentUsername();
@DLH.GetRequestHeaderValue("X-Authenticate-UserID")
``` 

beide liefern einen ``String`` zurück
